### PR TITLE
[SPARK-14315][SparkR]Add model persistence to GLMs

### PR DESCRIPTION
**Pull request by GayathriMurali**
## What changes were proposed in this pull request?

Add model persistence to GLMs in SparkR

Unit tests added

### CODE REVIEW

File: R/pkg/R/mllib.R [hunk 1]
```diff
@@ -83,6 +83,54 @@ setMethod("glm", signature(formula = "formula", family = "ANY", data = "SparkDat
...
+#' @export
+#' @examples
+#' \dontrun{
+#' df <- createDataFrame(sqlContext, infert)
+#' model <- glm(education ~ ., df)
+#' path <- "path/to/model"
+#' ml.save(model, path)
+#' }
+setMethod("ml.save", signature(object = "GeneralizedLinearRegressionModel", path = "character"),
+function(object, path, overwrite = FALSE) {
```
Comment (yanboliang) [inline R/pkg/R/mllib.R]: indent

File: R/pkg/R/mllib.R [hunk 2]
```diff
@@ -83,6 +83,54 @@ setMethod("glm", signature(formula = "formula", family = "ANY", data = "SparkDat
...
+#' @return a fitted MLlib model
+#' @rdname ml.load
+#' @name ml.load
+#' @export
+#' @examples
+#' \dontrun{
+#' path <- "path/to/model"
+#' model <- ml.load(path)
+#' }
+ml.load <- function(path) {
```
Comment (yanboliang) [inline R/pkg/R/mllib.R]: See my comments at #12680.

File: R/pkg/inst/tests/testthat/test_mllib.R [hunk 3]
```diff
@@ -126,6 +126,18 @@ test_that("glm summary", {
...
 
+  # Test model save/load
+  modelPath <- tempfile(pattern = "GLM", fileext = ".tmp")
+  ml.save(m, modelPath)
+  expect_error(ml.save(m, modelPath))
+  ml.save(m, modelPath, overwrite = TRUE)
+  m2 <- ml.load(modelPath)
+  s2 <- summary(m2)
+  expect_equal(s$rCoefficients, s2$rCoefficients)
+  expect_equal(s$rFeatures, s2$rFeatures)
```
Comment (yanboliang) [inline R/pkg/inst/tests/testthat/test_mllib.R]: I'd like to add more equality test for ```dispersion, deviance, null.deviance, df.null, null, residual, aic```.

**Comment by mengxr** [comment]
ok to test

**Comment by mengxr** [comment]
@GayathriMurali Please run `dev/lint-r` locally to check R style.

**Comment by mengxr** [comment]
@yanboliang Please `LGTM` if the changes look good to you:)

File: R/pkg/R/mllib.R [hunk 4]
```diff
@@ -406,6 +432,8 @@ ml.load <- function(path) {
   jobj <- callJStatic("org.apache.spark.ml.r.RWrappers", "load", path)
   if (isInstanceOf(jobj, "org.apache.spark.ml.r.NaiveBayesWrapper")) {
     return(new("NaiveBayesModel", jobj = jobj))
+  } else if (isInstanceOf(jobj, "org.apache.spark.ml.GeneralizedLinearRegressionWrapper")) {
```
Comment (yanboliang) [inline R/pkg/R/mllib.R]: Should be ```org.apache.spark.ml.r.GeneralizedLinearRegressionWrapper```.

Comment (GayathriMurali) [inline R/pkg/R/mllib.R]: For some reason local tests(R/run-tests.sh) is not capturing these failures. Let me fix this and submit the code

**Comment by yanboliang** [comment]
@GayathriMurali You should modify [```RWrappers.load```](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/r/RWrappers.scala#L44) at Scala side to make it support loading GLM wrapper. Thanks.

File: R/pkg/inst/tests/testthat/test_mllib.R [hunk 5]
```diff
@@ -71,7 +71,25 @@ test_that("glm and predict", {
...
 
-  # Test stats::predict is working
+  # Test model save/load
+  modelPath <- tempfile(pattern = "GLM", fileext = ".tmp")
+  ml.save(model, modelPath)
+  expect_error(ml.save(model, modelPath))
+  ml.save(model, modelPath, overwrite = TRUE)
+  m2 <- ml.load(modelPath)
+  s2 <- summary(m2)
+  expect_equal(s$rCoefficients, s2$rCoefficients)
```
Comment (yanboliang) [inline R/pkg/inst/tests/testthat/test_mllib.R]: There is no member variable ```rCoefficients``` in ```summary.GeneralizedLinearRegressionModel```, it should be ```coefficients```. Please refer L89-L98 in test_mllib.R, and the variables in the following line has the same issues.

Comment (yanboliang) [inline R/pkg/inst/tests/testthat/test_mllib.R]: And further more, we can have a separate test case named "glm save/load" which can test the equality of both the model and summary.

Comment (GayathriMurali) [inline R/pkg/inst/tests/testthat/test_mllib.R]: I agree. Should we do this test for all Gaussian, Poisson and Binomial family. I am assuming doing for either one of them should be sufficient?

**Pull request closed without merging**